storage-node/src/services/sync/acceptPendingObjects.ts

```diff
@@ -54,61 +59,81 @@ export class AcceptPendingObjectsService {
+  }
+
   private runWithInterval(api: ApiPromise, workerId: number, maxTxBatchSize: number, intervalMs: number) {
     const run = () => {
-      this.acceptPendingDataObjects(api, workerId, maxTxBatchSize)
-        .catch((err) => logger.error(`Failed to register pending data objects as accepted in runtime: ${err}`))
+      this.getPendingObjectsFromFolder()
+        .then((ids) => this.processPendingObjects(ids))
+        .then((objectsToAccept) => this.acceptPendingDataObjects(api, workerId, objectsToAccept, maxTxBatchSize))
+        .catch((err) => logger.error(`Error processing pending objects: ${err}`))
         .finally(() => setTimeout(run, intervalMs))
     }
     run()
   }
 
-  getPendingDataObject(objectId: string): [string, string] | undefined {
-    return this.pendingDataObjects.get(objectId)
-  }
-
-  private async loadPendingDataObjectsFromIDs(pendingIds: string[]): Promise<void> {
+  private async processPendingObjects(pendingIds: string[]): Promise<PendingObjectDetails> {
     const pendingDataObjects = await this.qnApi.getDataObjectsByIds(pendingIds)
 
-    const deletedObjects = _.differenceWith(pendingIds, pendingDataObjects, (id, dataObject) => dataObject.id === id)
+    // objects not found in the query node
+    const maybeDeletedObjects = _.differenceWith(
+      pendingIds,
+      pendingDataObjects,
+      (id, dataObject) => dataObject.id === id
+    )
 
-    // Remove stale objects from the pending directory
-    if (deletedObjects.length) {
-      logger.warn(`Found data objects deleted from runtime in Pending directory: ${deletedObjects.length}`)
-      await Promise.all(deletedObjects.map((id) => fsPromises.unlink(path.join(this.pendingDataObjectsDir, id))))
+    if (maybeDeletedObjects.length) {
+      // unlink the pending object once we determined it is gone!
+      logger.debug(`Found ${maybeDeletedObjects.length} stale pending objects in pending folder`)
+      // Things to consider.
+      // These may be very recently uploaded files, note the uploads api uses chain rpc to
+      // validate upload of object that was created to allow uploads for objects that were created
+      // as soon as possible.
+      // const qnState = await this.qnApi.getQueryNodeState()
+      // get this.api chain head
+      // get timestamp in chain head block.. are we uptodate?
+      // test api.isSyncing?
+      // It would be easier to have query-node store object deleted events.
+      // if we chose a different file name convention for the pending objects,
+      // we could store the bag-id and object id in the filename to make it easy to lookup in runtime state.
     }
 
-    await Promise.all(
+    const objectsToAccept: PendingObjectDetails = []
+
+    await Promise.allSettled(
       pendingDataObjects.map(async (dataObject) => {
-        const storageBucket = dataObject.storageBag.storageBuckets.find(({ id }) => this.uploadBuckets.includes(id))
+        // verify the ipfshash of the object matches against runtime!
```

Comment:
@zeeshanakram3 what are your thoughts about doing this additional check here?

Reply:
This check is already being [done](https://github.com/mnaamani/joystream/blob/3af466c526e37c40642f9045153f8377a5393739/storage-node/src/services/webApi/controllers/filesApi.ts#L139) after the file upload completes and before it is moved to the pending `dir`. So I think doing that would be redundant?

Reply:
It is of course redundant for objects that just came in through the http api, but not for files that might have been copied by operator directly. I'm thinking of scenario where operator maybe recovering their setup, moving files from other location after chaing pending folder path. Maybe somehow recovering "lost" objects from a backup/archive and placing them into the pending folder for processing..

Reply:
Yeah, in those cases it makes sense to have this validation. So maybe just add a comment, briefly explaining why are we doing this? In in which scenarios it would be beneficial? As you described above

Reply:
done in [0cdd273](https://github.com/Joystream/joystream/pull/5016/commits/0cdd2732bb3333040b16b714a081aeb46d8a8bc5)